### DISCUSSION TITLE
Trying to use kameo with tokio-tungstentie

### DISCUSSION THREAD

**Discussion by bananasov** [tqwewe/kameo#66] [Q&A] [answered]
Heyo! I've been trying kameo and so far I am absolutely loving the library, although i can't seem to wrap my brain around how to use the stream example with tokio-tungstenite websocket streams.

I'm trying to implement a client for a WebSocket system that sends messages to the client and receives them from a client, I want to be able to send the incoming messages from the stream to the person that created the actor and vice versa.

I've tried to just use [tokio_tungstenite::connect_async](https://docs.rs/tokio-tungstenite/latest/tokio_tungstenite/fn.connect_async.html) return value with [ActorRef::attach_stream](https://docs.rs/kameo/0.12.0/kameo/actor/struct.ActorRef.html#method.attach_stream) but that just resulted in errors I don't understand. 
```rs
let (ws_stream, _) = connect_async(&url).await.expect("Failed to connect");
let stream = Box::pin(ws_stream);
actor_ref.attach_stream(stream, "1st stream", "1st stream");
```

Does anyone have an idea? I am absolutely stumped.

**Comment by tqwewe** [comment, marked as answer]
Hi @bananasov, thanks for asking this! The errors returned are definitely quite obscure.

Essentially, in order to meet the requirements of `attach_stream`, your actor needs to implement `Message<StreamMessage<T, S, F>>`. [StreamMessage](https://docs.rs/kameo/latest/kameo/message/enum.StreamMessage.html) is a type provided by kameo, and is essentially just a wrapper type around the stream items. `S` and `F` can be any type you want for sending through context when the stream starts and finishes.

Looking into the [`Stream` impl of `WebSocketStream`](https://docs.rs/tokio-tungstenite/latest/tokio_tungstenite/struct.WebSocketStream.html#impl-Stream-for-WebSocketStream%3CT%3E), the `Item` type is `Result<Message, Error>`, which indicates your actor should have the following impl block:

```rust
use kameo::message::{Context, Message, StreamMessage};

impl
    Message<
        StreamMessage<
            Result<tungstenite::protocol::Message, tungstenite::error::Error>,
            &'static str,
            &'static str,
        >,
    > for MyActor
{
    type Reply = Result<(), tungstenite::error::Error>;

    async fn handle(&self, msg: StreamMessage, _ctx: Context<'_, Self, Self::Reply>) -> Self::Reply {
        match msg {
            StreamMessage::Next(ws_result) => {
                let ws_msg = ws_result?;
                // handle the websocket message
            }
            StreamMessage::Started(started_msg) => /* handle started */,
            StreamMessage::Finished(finished_msg) => /* handle finished */,
        }
    }
}
```

With this implementation, you should be able to call `actor_ref.attach_stream` using the stream, and your two `&'static str` items :)

Hopefully this clears it up and solves the errors you're having

**Reply by bananasov** [reply]
That clears it up by a lot! I'm wondering how the actor ref would receive messages from the stream though, without sending a message first

**Reply by tqwewe** [reply]
The attach stream method essentially spawns a dedicated tokio task which iterates the stream and forwards the items as messages to the actor.

https://github.com/tqwewe/kameo/blob/f2eaf24af18f7ded0ac9451c51d9102eb285badc/src/actor/actor_ref.rs#L521-L549

This is quite useful if you want to attach the stream in the background. However, if you're already in a running task lives the duration of the websocket connection, you can also manually message your actor:
```rust
while let Some(item) = stream.next().await? {
    actor_ref.tell(item).send().await?;
}
```

If your actors entire purpose is just for handling the websocket connection and will not be messaged in any part of your system (i.e. your only message impl is for `StreamMessage` and nothing else), then you may not even need an actor for this and instead simply handle the websocket connection in a tokio task.
